### ISSUE TITLE
Make it easier to train at a specific power output

### ISSUE THREAD

**Issue by gmar** [Sublimis/UrbanBiker#70]
For example if I want to do a Zone 2 ride at 150 Watts then you need a 3 second average power to be displayed in the UI. This lets you hold a specific power output value more easily (than watching the instantaneous power value jumping around). You would also want to see the instantaneous power value on the screen at the same time.

**Comment by Sublimis** [comment]
Hi, sorry for the late reply. We assume you're using a power meter sensor. If that's the case, than the value you see on the screen is in fact a 3 second average power. Can you please elaborate in more details the exact issue you're experiencing?

Note that power is by definition a derivative, and in fact it's a derivative of already noisy data, which further increases noise. Thus it's normal for power values to fluctuate. 3 second average is really not enough to smooth out all fluctuations.

Hope this helps. Thanks for the feedback!

**Comment by gmar** [comment]
Hi Sublimis,

Thanks for the response. There's no 'defect' per se that I'm referring to, it's more a "here's what I'm trying to accomplish - can the software find a way to help me do it".
When the Covid lock-down started, I bought a Tacx smart trainer and started using it with Zwift. Using ERG mode, it's very easy to work-out at a specific power output (because the trainer is constantly adjusting 'load' to keep you working at the same rate).
I bought a power meter for my bike - and was having trouble trying to perform the same workouts, in the same way - because I found it harder to get 'feedback' from the software about whether I was outputting too little, or too much power.

After getting interested in 'training with power', I've learnt a lot in recent months. So I now know that the power output shown within Zwift when ERG mode is enabled is very heavily 'shaped' to look much more 'stable' than it actually is. So I'm now aware that having a 'real power meter' produce the sort of feedback, you get in Zwift is unrealistic.

So while I'm now better at riding (on a real bike with a power meter) at a constant power output, it does seem like something that people want to be able to do (i.e. interval training) and so _maybe_ there is something the software can do to help people who are a bit new at it. 
I think there are a few ways to accomplish this (i.e. UI techniques) - having averaged power over 3 seconds seems the most simplistic one.
Anyway, if you think this is problem 'worth pursuing', I can follow up with some more ideas.

**Comment by zett42** [comment]
I think the current 3 second window is way too short for the purpose of riding at "constant" power. It requires one to continously look at the display and the spikes make it hard to get a feel for the average.

Suggestion: Make the averaging interval configurable, so users can experiment what works best for them for riding at constant power. I believe that something like 10 to 20 seconds could be sensible, but really one has to try this out for themselfs.

Another idea: Make it possible to configure desired min/max watts. Add a visual clue (e. g. change background color) when calculated average value is less than minimum or greater than maximum configured value.

**Comment by Sublimis** [comment]
Hi, did you have a chance to try this feature in the latest app release (v5.106)? What we did, is introduce a feature under the **Sensor properties** screen, below the "Data filtering" option, where one can adjust the amount of filtering inflicted on sensor data. The options range from 3 seconds to 30 seconds.

Thanks for the feedback!

![image](https://user-images.githubusercontent.com/3844660/93022832-7b1f7400-f5eb-11ea-8093-7fb7b1edede8.png)

**Comment by zett42** [comment]
Today I have set the interval to 10 s. Still, the displayed power value (watts) fluctuated like before, as often as once per second.
Also I noticed that sometimes a negative power value was displayed, it seemed to happen when coasting and even a few seconds after coasting, while already applying pressure to the pedals.

Is it possible that Urban Biker still calculates power from other sources, even when powermeter is attached? If so, how can I switch off calculated power and only use powermeter value?

**Comment by Sublimis** [comment]
> Is it possible that Urban Biker still calculates power from other sources, even when powermeter is attached? If so, how can I switch off calculated power and only use powermeter value?

No, this should not happen as long as the sensor is connected (which you can check at any time by looking at the "sensors" icon in the statusbar -- it should be green throughout the ride). Values from the sensor are used whenever available. Only when the sensor gets disconnected Urban Biker falls back to calculated values, and this behavior cannot be disabled.

> Today I have set the interval to 10 s. Still, the displayed power value (watts) fluctuated like before, as often as once per second.

Can you please try with a 30 seconds filter interval, and by using the latest v5.107? We could try implementing some other kind of filter in one of the next versions, if you are not satisfied with the results.

Thanks for the feedback!

**Comment by zett42** [comment]
Using latest BETA with 30 s filter interval, the power value appears to be more stable. Though for my taste the lag was a little bit too long, so I have set the interval to 15 s. It worked quite good when I did a regeneration ride yesterday where I payed attention to keep my power output low.

Question: Does the filter also affect recorded power values?
